playbook/queue_consumer.go

```diff
@@ -0,0 +1,159 @@
+package playbook
+
+import (
+	"encoding/json"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/flanksource/commons/logger"
+	"github.com/flanksource/duty/models"
+	"github.com/flanksource/incident-commander/api"
+	v1 "github.com/flanksource/incident-commander/api/v1"
+	"github.com/flanksource/incident-commander/db"
+	"github.com/flanksource/incident-commander/utils"
+	"github.com/google/uuid"
+	"github.com/jackc/pgx/v5/pgxpool"
+	"gorm.io/gorm"
+)
+
+type queueConsumer struct {
+	pool                           *pgxpool.Pool
+	db                             *gorm.DB
+	tickInterval                   time.Duration
+	dbReconnectMaxDuration         time.Duration
+	dbReconnectBackoffBaseDuration time.Duration
+
+	// registry stores the list of playbook run IDs
+	// that are currently being executed.
+	registry sync.Map
+}
+
+func NewQueueConsumer(db *gorm.DB, pool *pgxpool.Pool) *queueConsumer {
+	return &queueConsumer{
+		db:                             db,
+		pool:                           pool,
+		tickInterval:                   time.Minute,
+		dbReconnectMaxDuration:         time.Minute,
+		dbReconnectBackoffBaseDuration: time.Second,
+		registry:                       sync.Map{},
+	}
+}
+
+func (t *queueConsumer) Listen() error {
+	pgNotify := make(chan string)
+	go utils.ListenToPostgresNotify(t.pool, "playbook_run_updates", t.dbReconnectMaxDuration, t.dbReconnectBackoffBaseDuration, pgNotify)
+
+	pgNotifyPlaybookSpecApprovalUpdated := make(chan string)
+	go utils.ListenToPostgresNotify(t.pool, "playbook_spec_approval_updated", t.dbReconnectMaxDuration, t.dbReconnectBackoffBaseDuration, pgNotifyPlaybookSpecApprovalUpdated)
+
+	pgNotifyPlaybookApprovalsInserted := make(chan string)
+	go utils.ListenToPostgresNotify(t.pool, "playbook_approval_inserted", t.dbReconnectMaxDuration, t.dbReconnectBackoffBaseDuration, pgNotifyPlaybookApprovalsInserted)
+
+	ctx := api.NewContext(t.db, nil)
+	for {
+		select {
+		case <-pgNotify:
+			if err := t.consumeAll(ctx); err != nil {
+				logger.Errorf("%v", err)
+			}
+
+		case id := <-pgNotifyPlaybookSpecApprovalUpdated:
+			if err := t.onPlaybookSpecApprovalUpdated(ctx, id); err != nil {
+				logger.Errorf("%v", err)
+			}
+
+		case id := <-pgNotifyPlaybookApprovalsInserted:
+			if err := t.onPlaybookRunNewApproval(ctx, id); err != nil {
+				logger.Errorf("%v", err)
+			}
+
+		case <-time.After(t.tickInterval):
+			if err := t.consumeAll(ctx); err != nil {
+				logger.Errorf("%v", err)
+			}
+		}
+	}
+}
+
+func (t *queueConsumer) onPlaybookSpecApprovalUpdated(ctx *api.Context, playbookID string) error {
```

Comment:
These functions are all in the playbook package so no need to duplicate prefix
```suggestion
func (t *queueConsumer) onApprovalUpdated(ctx *api.Context, playbookID string) error {
```